pkg/backup/backup.go

```diff
@@ -174,94 +175,137 @@ func (b *Backup) Run() {
 	}
 }
 
+// saveSnap saves the latest snapshot if the given current revision is greater lastSnapRev.
 func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
-	podList, err := b.kclient.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+	start := time.Now()
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	rc, etcdVersion, rev, err := GetSnap(ctx, b.kclient, b.etcdTLSConfig, b.namespace, b.clusterName)
+	defer cancel()
+	defer rc.Close()
 	if err != nil {
-		return lastSnapRev, err
+		return lastSnapRev, fmt.Errorf("failed to get snapshot: (%v)", err)
 	}
 
-	var pods []*v1.Pod
-	for i := range podList.Items {
-		pod := &podList.Items[i]
-		if pod.Status.Phase == v1.PodRunning {
-			pods = append(pods, pod)
-		}
+	if rev <= lastSnapRev {
+		logrus.Info("skipped creating new backup: no change since last time")
+		return lastSnapRev, nil
 	}
 
-	if len(pods) == 0 {
-		msg := "no running etcd pods found"
-		logrus.Warning(msg)
-		return lastSnapRev, fmt.Errorf(msg)
-	}
-	member, rev := getMemberWithMaxRev(pods, b.etcdTLSConfig, b.selfHosted)
-	if member == nil {
-		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+	n, err := WriteSnap(etcdVersion, rev, b.be, rc)
+	if err != nil {
+		return 0, fmt.Errorf("failed to write snapshot to backend: (%v)", err)
 	}
 
-	if rev <= lastSnapRev {
-		logrus.Info("skipped creating new backup: no change since last time")
-		return lastSnapRev, nil
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
+	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
 	}
 
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+	return 0, nil
+}
+
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, rc io.ReadCloser) (int64, error) {
+	n, err := be.Save(version, rev, rc)
+	if err != nil {
+		return 0, err
 	}
-	return rev, nil
+	return n, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
```

Comment:
document that ctx should be closed when returned io.ReadCloser is done reading.

Reply:
Discussed with @fanminshi . My fault I didn't understand Snapshot() api good enough..
Need to refactor this part further to avoid background goroutine.